woocommerce/google-listings-and-ads
# Shipping rates and shipping times: Add and edit modals - validation logic

### Changes proposed in this Pull Request:

Closes #552 .

This PR adds validation for add and edit shipping rates and shipping times modals in Setup MC and edit free listings pages.

#### Technical Note:

##### AppInputPriceControl and AppInputNumberControl

I have refactored the `AppInputPriceControl` with a new component `AppInputNumberControl` which is used to handle the shipping time ("number of days", which should be an integer number) input. I have also created a number of functions and hooks to work with the integer number (e.g. `useStoreNumberSettings`, `useNumberFormat`).

##### Currency and Number

The relationship between currency and number in the app can be quite complicated and deserves a lil bit more explanation.

In WooCommerce settings, we have currency options (available in https://gla1.test/wp-admin/admin.php?page=wc-settings&tab=general), and we can get the currency settings using `useStoreCurrency` hook. This is the  _WooCommerce business domain_.

To make things work with general integer number, I came up with a `useStoreNumberSettings` which wraps the `useStoreCurrency` and passes in `{ precision: 0 }`. At first glance, this can be a bit funny because, in general, "currency is a type of number" (which means it should be currency wrapping/inheriting number), but the other way around is not true. However, because of the WooCommerce settings above, I think this approach is sensible and it would automatically respect the `decimalSeparator` and `thousandSeparator` specified in the WC currency options.

The `AppInputNumberControl` is a general integer number input control. It makes use of the `useStoreNumberSettings` above. We can customize the number precision via the `numberSettings` props.

The `AppInputPriceControl` is a more specific type of `AppInputNumberControl` where it passes in the store currency settings and displays store currency code in the input suffix.

The following illustrates the component-data flow:

```
(UI) AppInputPriceControl --wraps--> AppInputNumberControl --use--> useStoreNumberSettings --use--> useStoreCurrency (business domain)
```

##### Change shipping rate and time to number in wp-data reducer

In wp-data, upon receiving shipping rates and times, I converted them into numbers and store them in wp-data. This would make them easier to work with the input controls and validation.

##### Pre-filled shipping rate and time in Setup MC

In the Setup MC flow, we have the pre-filled shipping rate in `shipping-rate-setup/index.js ` and shipping time in `shipping-time-setup/index.js`. Prior to this PR, the rate and time are set to empty string `''`.

In this PR, I have changed them to be `null` to indicate absence of value. Note that we should not make them `0` because that would show up as `0` in the front-end but it is not saved yet in the back-end.

##### Shipping rates and times inconsistent naming technical debt

We are using a few different terms for the same thing in different forms, some examples:

- `countries`, `countryCodes`
- `rate`, `price`

Because of the data flow logics involved, making them consistent to share validation logic requires a bit more effort and is out of scope of this PR. 

In the future, when we make Setup MC and Edit Free Listings share same component (as noted in https://github.com/woocommerce/google-listings-and-ads/issues/288), it would be easier to share validation logic.

### Screenshots:

Demo video with my voice: (https://d.pr/v/aCYsSd)

https://user-images.githubusercontent.com/417342/119716519-c8045b80-be97-11eb-9345-961750b971c5.mov

### Detailed test instructions:

1. Open https://gla1.test/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fsetup-mc
2. Proceed to Step 3.
3. In shipping rates section, click on edit button to bring up the edit modal. When there are no country specified, no rate specified, or rate is below 0, the save button should be disabled.
4. Click on add button to bring up the add modal. When there are no country specified, no rate specified, or rate is below 0, the save button should be disabled.
5. In the shipping times section, check for the same thing with the edit modal and add modal.
6. When Setup MC is done, go to Edit Free Listings and check for the same thing as per Step 3, 4, and 5 above.

### Changelog Note:

Add validation for shipping rates and shipping times modals.
